File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/flutterobject/BaseNativeObject.java
package com.valentingrigorean.arcgis_maps_flutter.flutterobject;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import io.flutter.plugin.common.MethodChannel;

public abstract class BaseNativeObject<T> implements NativeObject, NativeMessageSink {

    private final T nativeObject;
    private final String objectId;
    private final NativeHandler[] nativeHandlers;

    private NativeObjectMessageSink nativeObjectMessageSink;
    private NativeMessageSink messageSink;

    private boolean isDisposed = false;

    protected BaseNativeObject(String objectId, T nativeObject, NativeHandler[] nativeHandlers) {
        this.objectId = objectId;
        this.nativeObject = nativeObject;
        this.nativeHandlers = nativeHandlers;

        for (NativeHandler nativeHandler : nativeHandlers) {
            nativeHandler.setMessageSink(this);
        }
    }

    @Override
    protected void finalize() throws Throwable {
        dispose();
        super.finalize();
    }

    public T getNativeObject() {
        return nativeObject;
    }

    @Override
    public final void send(@NonNull String method, @Nullable Object args) {
        if (messageSink == null || isDisposed) {
            return;
        }
        nativeObjectMessageSink.send(method, args);
    }

    @Override
    public String getObjectId() {
        return objectId;
    }

    @Override
    public void setMessageSink(@Nullable NativeMessageSink messageSink) {
        this.messageSink = messageSink;
        if (messageSink != null) {
            nativeObjectMessageSink = new NativeObjectMessageSink(objectId, messageSink);
        } else {
            nativeObjectMessageSink = null;
        }
    }

    @Override
    public final void dispose() {
        if (isDisposed) {
            return;
        }
        isDisposed = true;
        disposeInternal();

        for (NativeHandler nativeHandler : nativeHandlers) {
            nativeHandler.setMessageSink(null);
            nativeHandler.dispose();
        }
    }

    @Override
    public void onMethodCall(@NonNull String method, @Nullable Object args, @NonNull MethodChannel.Result result) {
        for (NativeHandler nativeHandler : nativeHandlers) {
            if (nativeHandler.onMethodCall(method, args, result)) {
                return;
            }
        }
        result.notImplemented();
    }

    protected void disposeInternal() {

    }

    protected NativeMessageSink getMessageSink() {
        return messageSink;
    }

    protected final NativeObjectStorage getNativeObjectStorage() {
        return NativeObjectStorage.getInstance();
    }
}

File: android/src/main/kotlin/com/valentingrigorean/arcgis_maps_flutter/flutterobject/BaseNativeObject.kt
package com.valentingrigorean.arcgis_maps_flutter.flutterobject

import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job

abstract class BaseNativeObject<T> protected constructor(
    final override val objectId: String,
    val nativeObject: T,
    private val nativeHandlers: Array<NativeHandler>
) : NativeObject, NativeMessageSink {
    private val messageSinkImpl = NativeObjectMessageSinkImpl(objectId)
    private var isDisposed = false
    private val coroutineJob = Job()
    protected val scope = CoroutineScope(Dispatchers.Main + coroutineJob)

    init {
        for (nativeHandler in nativeHandlers) {
            nativeHandler.setMessageSink(messageSinkImpl)
        }
    }

    override fun send(method: String, args: Any?) {
        if (isDisposed) {
            return
        }
        messageSinkImpl.send(method, args)
    }

    override fun setMessageSink(messageSink: NativeMessageSink?) {
        messageSinkImpl.setMessageSink(messageSink)
    }

    override fun dispose() {
        if (isDisposed) {
            return
        }
        isDisposed = true
        coroutineJob.cancel()
        for (nativeHandler in nativeHandlers) {
            nativeHandler.setMessageSink(null)
            nativeHandler.dispose()
        }
    }

    override fun onMethodCall(method: String, args: Any?, result: MethodChannel.Result) {
        for (nativeHandler in nativeHandlers) {
            if (nativeHandler.onMethodCall(method, args, result)) {
                return
            }
        }
        result.notImplemented()
    }

    protected fun getMessageSink(): NativeMessageSink {
        return messageSinkImpl
    }

    protected val nativeObjectStorage: NativeObjectStorage
        get() = NativeObjectStorage.instance
}

private class NativeObjectMessageSinkImpl(
    private val objectId: String,
    private var messageSink: NativeMessageSink? = null,
) : NativeMessageSink {

    private val data = HashMap<String, Any?>(3)

    override fun send(method: String, args: Any?) {
       if(messageSink == null){
           return
       }
        data.clear()
        data["objectId"] = objectId
        data["method"] = method
        data["arguments"] = args
        messageSink!!.send("messageNativeObject", data)
    }

    fun setMessageSink(messageSink: NativeMessageSink?) {
        this.messageSink = messageSink
    }
}